File: C++_Version/String/680Valid_Palindrome_II.cpp
//Given a non - empty string s, you may delete at most one character.Judge whether you can make it a palindrome.
//
//Example 1:
//Input: "aba"
//	Output : True
//	Example 2 :
//	Input : "abca"
//	Output : True
//	Explanation : You could delete the character 'c'.
//	Note :
//	The string will only contain lowercase characters a - z.The maximum length of the string is 50000.

#include <vector>
#include <iostream>
#include <algorithm>
#include <cstdlib>
#include <functional>
#include <string>
#include <unordered_map>
#include <map>
#include <set>
#include <unordered_set>
using namespace std;

class Solution {
public:
	bool validPalindrome(string s)
	{
		int s_length = s.length();

		int low = 0;
		int high = s_length - 1;
		bool flag = true;


		while (low <= high)
		{
			if (s[low] != s[high])
			{
				if (!flag)
					return false;

				flag = false;
				

				if (s[low] == s[high - 1] && s[low + 1] != s[high])
					high--;
				else if (s[low + 1] == s[high] && s[low] != s[high - 1])
					low++;
				else if (s[low + 1] != s[high] && s[low] != s[high - 1])
					return false;
				else
					return validPalindrome_helper(s, low + 1, high) || validPalindrome_helper(s, low, high - 1);
			}
			else
			{
				low++;
				high--;
			}
		}

		return true;
	}

private:
	bool validPalindrome_helper(string s, int low, int high)
	{
		while ( low <= high)
		{
			if (s[low] != s[high])
				return false;
			
			low++;
			high--;
		}

		return true;
	}
};

int main()
{
	Solution sol1;

	string s1("ebcbbececabbacecbbcbe");

	bool receive = sol1.validPalindrome(s1);

	system("pause");

	return 0;
}

File: Java_Version/680_Valid_Palindrome_II.java
/*
  Given a non-empty string s, you may delete at most one character. Judge
whether you can make it a palindrome.

Example 1:
        Input: "aba"
        Output: True

Example 2:
        Input: "abca"
        Output: True

        Explanation: You could delete the character 'c'.

Note:
  The string will only contain lowercase characters a-z. The maximum length of
the string is 50000.
*/

class Solution {

    public static void main(String[] args) {

        String s = "abcda";

        System.out.println(new Solution().validPalindrome(s));
    }

    public boolean validPalindrome(String s) {

        if (s == null || s.length() == 0) {
            return true;
        }

        return validPalindromeHelper(s, 0, s.length() - 1);
    }

    private boolean validPalindromeHelper(String s, int startIndex, int endIndex) {

        if (startIndex >= endIndex) {
            return true;
        }

        if (s.charAt(startIndex) == s.charAt(endIndex)) {
            return validPalindromeHelper(s, startIndex + 1, endIndex - 1);
        } else {
            return ifStrictlyPalindrome(s, startIndex, endIndex - 1)
                    || ifStrictlyPalindrome(s, startIndex + 1, endIndex);
        }
    }

    private boolean ifStrictlyPalindrome(String s, int startIndex, int endIndex) {

        while (startIndex < endIndex) {
            if (s.charAt(startIndex) != s.charAt(endIndex)) {
                return false;
            }
            startIndex++;
            endIndex--;
        }

        return true;
    }
}